### PR TITLE
Fix getPedMoveState returns incorrect move states

### PR DESCRIPTION
**Pull request by Unde-R**
This should fix the following issues:
#1598 
#1594 
#1590

### CODE REVIEW

**Comment by Lpsd** [comment]
Thanks for your PR. I'm sure this will be fine but I'll need to do some extensive testing later just to make sure it doesn't unintentionally break anything else. Good work though!

**Comment by Unde-R** [comment]
> Thanks for your PR. I'm sure this will be fine but I'll need to do some extensive testing later just to make sure it doesn't unintentionally break anything else. Good work though!

Thanks mate

**Comment by Unde-R** [comment]
I tested a code to get groundposition and compared it with player's Z position in the areas which return "fall" and here is the results ( Z - groundposition ) :
1,00000095367
1.00000095367
1.00000095368

**Comment by qaisjp** [comment]
thanks for submitting this. quick question for future documentation / reproduction purposes, please can you provide the coordinates you tested with? thank you :)

**Comment by Unde-R** [comment]
> thanks for submitting this. quick question for future documentation / reproduction purposes, please can you provide the coordinates you tested with? thank you :)

sure one moment

**Comment by Unde-R** [comment]
```
Ground Position: 14.250931739807
Player Position: 2221.7590332031, -1653.4652099609, 15.250932693481
-------------
Ground Position: 14.252799987793
Player Position: 2221.814453125, -1653.5106201172, 15.252800941467
-------------
Ground Position: 14.287864685059
Player Position: 2223.142578125, -1653.3934326172, 15.287865638733
```

File: Client/mods/deathmatch/logic/CClientPed.cpp [hunk 1]
```diff
@@ -4626,7 +4639,7 @@ bool CClientPed::IsOnGround()
     CVector vecPosition;
     GetPosition(vecPosition);
     float fGroundLevel = static_cast<float>(g_pGame->GetWorld()->FindGroundZFor3DPosition(&vecPosition));
-    return (vecPosition.fZ > fGroundLevel && (vecPosition.fZ - fGroundLevel) <= 1.0f);
+    return (vecPosition.fZ > fGroundLevel && (vecPosition.fZ - fGroundLevel) <= 1.01f);
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientPed.cpp]:
```suggestion

    // We don't care if the ped is underground
    if ((vecPosition.fZ - fGroundLevel) <= FLOAT_EPSILON)
        return false;
        
    // Check that (zPos - groundLevel) <= 1, but also account for epsilon
    return (vecPosition.fZ - fGroundLevel - 1.0f) <= FLOAT_EPSILON;
```

Does this do the same thing?

Comment (Unde-R) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: let me test

Comment (Unde-R) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: since FLOAT_EPSILON expands to 0.0001f it should work, gonna test it rn.

Comment (qaisjp) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: My suggested code seems theoretically incorrect (even if it might happen to work).

Looking at https://stackoverflow.com/a/77735/1517394:

> It all depends on context.
>
> I have spent a long time tracing a bugs in a system that presumed `a==b` if `|a-b|<epsilon`. The underlying problems were: [...]

And https://stackoverflow.com/a/253874/1517394:

> The following definitions are from [The art of computer programming by Knuth](http://books.google.ca/books?ei=KRkLSYLZEJ7ktQOX3-ChBQ&id=T89QAAAAMAAJ&dq=%22essentially+equal%22+%22approximately+equal%22+knuth&q=%22essentially+equal%22+%22approximately+equal%22+%22definitely+less+than%22&pgis=1#search):
> 
> ```cpp
> bool approximatelyEqual(float a, float b, float epsilon)
> {
>     return fabs(a - b) <= ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
> }
> 
> bool essentiallyEqual(float a, float b, float epsilon)
> {
>     return fabs(a - b) <= ( (fabs(a) > fabs(b) ? fabs(b) : fabs(a)) * epsilon);
> }
> 
> bool definitelyGreaterThan(float a, float b, float epsilon)
> {
>     return (a - b) > ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
> }
> 
> bool definitelyLessThan(float a, float b, float epsilon)
> {
>     return (b - a) > ( (fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
> }
> ```

Comment (Unde-R) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: Yeah probably, well we can lower the "1.01f" more according to the values that I got above, "1.01f" is already alot

Comment (Einheit-101) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: Why not use 1.001 instead of 1.01? 0.001 is very very low and no one will notice an issue with this, pretty sure

Comment (Unde-R) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: yeah I thought about it, just waiting for lpsd to test it

**Comment by Unde-R** [comment]
I added `MOVEMENTSTATE_ROLL` because It was declared but no code for It.

- [x] crouch-roll
- [x] crawl-roll

Result: [`https://pastebin.com/vz4zQb3D`](https://pastebin.com/vz4zQb3D)

**Comment by Einheit-101** [comment]
This is absolutely amazing!

**Comment by Unde-R** [comment]
Looks like this issue should be fixed too #534

**Review by qaisjp — CHANGES_REQUESTED** [review]
I'd like to see more correct handling, via `FLOAT_EPSILON`. (Assuming that this is actually the correct way to handle it.)

**Comment by Unde-R** [comment]
Thanks for the help!
Tested It's working fine now.

**Comment by qaisjp** [comment]
The suggested code I provided is not correct. Please read https://github.com/multitheftauto/mtasa-blue/pull/1603#discussion_r467206465

**Comment by Unde-R** [comment]
> The suggested code I provided is not correct. Please read [#1603 (comment)](https://github.com/multitheftauto/mtasa-blue/pull/1603#discussion_r467206465)

`return (vecPosition.fZ > fGroundLevel && (vecPosition.fZ - fGroundLevel - 1.0f) <=
                                                 ((fabs(vecPosition.fZ) < fabs(fGroundLevel) ? fabs(fGroundLevel) : fabs(vecPosition.fZ)) * FLOAT_EPSILON));` Thoughts on this?

**Comment by qaisjp** [comment]
Almost there! Remember, there's two checks here. One that checks `vecPosition.fZ > fGroundLevel`, and another that checks `(vecPosition.fZ - fGroundLevel) <= 1.0f`.

I think this should be split up into two if-s, with correct float comparisons.

To be honest, the code is a bit hard to read, so perhaps you should add helpers like [these](https://swift.im/git/swift-contrib/diff/Swiften/Base?id=c8ced8597ab28b928862d21554c0a68c238b1165&context=4)* to [`Utils.cpp`](https://github.com/multitheftauto/mtasa-blue/blob/master/Shared/mods/deathmatch/logic/Utils.cpp) or [`Utils.h`](https://github.com/multitheftauto/mtasa-blue/blob/master/Shared/mods/deathmatch/logic/Utils.h).

Then you can do something like this:

```cpp
    // We don't care if the ped is underground
    if (definitelyGreaterThan(vecPosition.fZ, fGroundLevel))
        return false;
        
    // Check (vecPosition.fZ - fGroundLevel) <= 1.0f
    return definitelyLessThan(vecPosition.fZ - fGroundLevel, 1.0f) || essentiallyEqual(vecPosition.fZ - fGroundLevel, 1.0f);
```

I'm not sure if this best practice, but it's the best thing I can come up, and I feel like it's better than just adding 0.001 and calling it a day.


[*] this link is from "Swift", but it is actually some chat software, and not the Swift Programming Language

**Comment by Unde-R** [comment]
> ```c++
>     // We don't care if the ped is underground
>     if (definitelyGreaterThan(vecPosition.fZ, fGroundLevel))
>         return false;
>         
>     // Check (vecPosition.fZ - fGroundLevel) <= 1.0f
>     return definitelyLessThan(vecPosition.fZ - fGroundLevel, 1.0f) || essentiallyEqual(vecPosition.fZ - fGroundLevel, 1.0f);
> ```
> 
I tried this ^^ but somehow It's not working properly.

**Comment by Unde-R** [comment]
> 
```c++
// We don't care if the ped is underground
    if (definitelyGreaterThan(vecPosition.fZ, fGroundLevel))
        return false;
```
>
Looks like It's because of this.

**Comment by Pirulax** [comment]
Do this (like in the CVector class): 
```cpp
abs(vec.z) - abs(groundposz) >= FLOAT_EPSILON
```
This is totally acceptable, as `float epsilon` is a thing in CS

**Comment by qaisjp** [comment]
```cpp
    // We don't care if the ped is underground
    if (definitelyGreaterThan(vecPosition.fZ, fGroundLevel))
        return false;
```

The code I suggested had a small mistake.

It doesn't work because we're returning `false` if our z-position is greater than the ground level. That does not match up to the comment that says "we don't care if the ped is underground"?

It should return false if the ped is underground. But we're not checking if the ped is underground correctly.

you should use `definitelyLessThan` instead :)

**Comment by Unde-R** [comment]
Yeah I'll update it soon.
What about crouch-roll should I keep it here or make another PR for it?

**Comment by qaisjp** [comment]
> What about crouch-roll should I make another PR for it?

Good idea

**Comment by Unde-R** [comment]
> Good idea

I'll open a PR rn

**Review by ghost — DISMISSED** [review]

**Comment by Einheit-101** [comment]
This is tested a lot and seems to be fine, why is this not merged yet?

**Comment by patrikjuvonen** [comment]
> This is tested a lot and seems to be fine, why is this not merged yet?

No accepted reviews, and no testing beyond the stale ones has been made. I am interested to hear if this actually resolves the three linked issues and how does it do that since it is not apparent from the description. This leaves me to have to thoroughly delve into the code to figure out whats going on, how to test it and how to try to break it. None of which I have time to do right now personally.

**Comment by Einheit-101** [comment]
I could test functionality with a working build if i can get one. My resource uses these functions.

**Comment by ghost** [comment]
@patrikjuvonen 

Tested on nightly (Multi Theft Auto v1.5.9-release-21251 - but the version shouldn't matter, since it exists on stable as well)

#1598 - [fixed build](https://www.youtube.com/watch?v=Pun1AZwban0) & [nightly build](https://www.youtube.com/watch?v=vAMUaDXrHvY)
#1594 - [fixed build](https://www.youtube.com/watch?v=oqmr-yYHdnk) & [nightly build](https://www.youtube.com/watch?v=ueNFJcby9xM)
#1590 - [fixed build](https://www.youtube.com/watch?v=jDmoqu4tnaM) & [nightly build](https://www.youtube.com/watch?v=7x_N6ldlvvw)

Besides those locations, i've tested whether it happens in the other [places](https://imgur.com/a/k9y8UEc), which haven't been mentioned. And it seems that fix works well.

**Comment by patrikjuvonen** [comment]
> @patrikjuvonen
> 
> Tested on nightly (Multi Theft Auto v1.5.9-release-21251 - but the version shouldn't matter, since it exists on stable as well)
> 
> #1598 - [fixed build](https://www.youtube.com/watch?v=Pun1AZwban0) & [nightly build](https://www.youtube.com/watch?v=ueNFJcby9xM) #1594 - [fixed build](https://www.youtube.com/watch?v=oqmr-yYHdnk) & [nightly build](https://www.youtube.com/watch?v=ueNFJcby9xM) #1590 - [fixed build](https://www.youtube.com/watch?v=jDmoqu4tnaM) & [nightly build](https://www.youtube.com/watch?v=7x_N6ldlvvw)
> 
> Besides those locations, i've tested whether it happens in the other [places](https://imgur.com/a/k9y8UEc), which haven't been mentioned. And it seems that fix works well.

I think you shared the wrong nightly build video for 1598, it's the same as 1594. Tests should be equal.

**Comment by patrikjuvonen** [comment]
I don't think 1594 is fixed by this PR (based on the video above). Why is it returning false for move state when the move state is supposed to be something specific? It will still be causing the same headache.

**Comment by Unde-R** [comment]
> I don't think 1594 is fixed by this PR (based on the video above). Why is it returning false for move state when the move state is supposed to be something specific? It will still be causing the same headache.

because of this
https://github.com/multitheftauto/mtasa-blue/pull/1603/files#diff-25862afc0e3088d4619017944046a0d75c9a75a9ffbdb666216bb3a9478453fdR2381

**Comment by Unde-R** [comment]
> I don't think 1594 is fixed by this PR (based on the video above). Why is it returning false for move state when the move state is supposed to be something specific? It will still be causing the same headache.

basically in the video of the fixed build, he was standing at first then he's trying to enter the vehicle thats why its showing false, because he's entering the vehicle
```c++
        else if (IsEnteringVehicle())
            return MOVEMENTSTATE_UNKNOWN;
```

**Comment by patrikjuvonen** [comment]
Exactly. I don't understand why it has been specified as unknown. If the ped is jogging to enter a vehicle, shouldn't it return jogging?

**Comment by ghost** [comment]
@patrikjuvonen i guess qaisjp [agreed upon making it false](https://github.com/multitheftauto/mtasa-blue/issues/1594#issuecomment-669576930)

He also suggested it:

```
Vehicle-related movement states

Note that getPedMoveState documents what happens when the player is in a vehicle:

    Returns a string indicating the ped's move state, or false if the ped is not streamed in, the movement type is unknown, the ped is in a vehicle or the ped is invalid.

Basically, I'm not 100% sure whether enter_vehicle and leave_vehicle fits well with the rest of these move states. What do you think?
```

**Comment by ghost** [comment]
> I think you shared the wrong nightly build video for 1598, it's the same as 1594. Tests should be equal.

Oh sorry, you're right, i added wrong url. Fixed it (and in previous post as well) - https://www.youtube.com/watch?v=vAMUaDXrHvY

It might be hard to see, but it changes to "fall" once in a time (0:12 on this screen)
obraz

**Comment by patrikjuvonen** [comment]
> > I think you shared the wrong nightly build video for 1598, it's the same as 1594. Tests should be equal.
> 
> Oh sorry, you're right, i added wrong url. Fixed it (and in previous post as well) - https://www.youtube.com/watch?v=vAMUaDXrHvY
> 
> It might be hard to see, but it changes to "fall" once in a time (0:12 on this screen) obraz

Thanks.

> @patrikjuvonen i guess qaisjp [agreed upon making it false](https://github.com/multitheftauto/mtasa-blue/issues/1594#issuecomment-669576930)
> 
> He also suggested it:
> 
> ```
> Vehicle-related movement states
> 
> Note that getPedMoveState documents what happens when the player is in a vehicle:
> 
>     Returns a string indicating the ped's move state, or false if the ped is not streamed in, the movement type is unknown, the ped is in a vehicle or the ped is invalid.
> 
> Basically, I'm not 100% sure whether enter_vehicle and leave_vehicle fits well with the rest of these move states. What do you think?
> ```

That is not his suggestion, the quote is what is documented in Wiki under https://wiki.multitheftauto.com/wiki/GetPedMoveState#Returns

In #1594 the ped is valid, streamed in, is not in a vehicle and I would expect the movement state to be known as well, since it returned non-false before. So I just don't understand why we're now downgrading it to return `false`. If it's now a limitation for some reason, then that's fine, but it's not fixing the issue. Alternatively we can close that issue and create a new one about an unknown state when entering a vehicle, because I believe that still needs to be worked out.

**Comment by Unde-R** [comment]
> > > I think you shared the wrong nightly build video for 1598, it's the same as 1594. Tests should be equal.
> > 
> > 
> > Oh sorry, you're right, i added wrong url. Fixed it (and in previous post as well) - https://www.youtube.com/watch?v=vAMUaDXrHvY
> > It might be hard to see, but it changes to "fall" once in a time (0:12 on this screen) obraz
> 
> Thanks.
> 
> > @patrikjuvonen i guess qaisjp [agreed upon making it false](https://github.com/multitheftauto/mtasa-blue/issues/1594#issuecomment-669576930)
> > He also suggested it:
> > ```
> > Vehicle-related movement states
> > 
> > Note that getPedMoveState documents what happens when the player is in a vehicle:
> > 
> >     Returns a string indicating the ped's move state, or false if the ped is not streamed in, the movement type is unknown, the ped is in a vehicle or the ped is invalid.
> > 
> > Basically, I'm not 100% sure whether enter_vehicle and leave_vehicle fits well with the rest of these move states. What do you think?
> > ```
> 
> That is not his suggestion, the quote is what is documented in Wiki under https://wiki.multitheftauto.com/wiki/GetPedMoveState#Returns
> 
> In #1594 the ped is valid, streamed in, is not in a vehicle and I would expect the movement state to be known as well, since it returned non-false before. So I just don't understand why we're now downgrading it to return `false`. If it's now a limitation for some reason, then that's fine, but it's not fixing the issue. Alternatively we can close that issue and create a new one about an unknown state when entering a vehicle, because I believe that still needs to be worked out.

before it did return fall when entering or leaving I assume

**Comment by patrikjuvonen** [comment]
> before it did return fall when entering or leaving I assume

Before this patch was initially introduced, that is (2020)? Because it's correct in this video srslyyyy posted: https://youtu.be/ueNFJcby9xM

**Comment by patrikjuvonen** [comment]
I think we might be talking about two different things. There's ped moving to the "door"/entrypoint and actually physically entering the vehicle.

The actual entering animation is ok to be false.

For the moving to the "door"/entrypoint it shouldn't be false.

**Comment by Unde-R** [comment]
We can do when he's entering the vehicle then we get the task name and see?

**Comment by Unde-R** [comment]
"TASK_GROUP_ENTER_CAR" what about this?
but idk about group task

**Comment by Einheit-101** [comment]
I just tested this. It works better than before, but 3 issues:
- Move state returns "fall" while wielding a jetpack (Maybe not an issue, but noteworthy)
- Move state returns false while trying to enter / exit a vehicle and the Ped is still running towards the vehicle
- Move state returns "fall" while standing on certain dynamic objects like this car wreck: https://imgur.com/a/zEB0Ljx

The last issue is the only problematic one because the other two can be fixed with scripting. It looks like all dynamic objects cause "fall".

File: Client/mods/deathmatch/logic/CClientPed.cpp [hunk 2]
```diff
@@ -2377,6 +2378,9 @@ eMovementState CClientPed::GetMovementState()
         else if (strcmp(szComplexTaskName, "TASK_COMPLEX_JUMP") == 0)
             return MOVEMENTSTATE_JUMP;
 
+        else if (IsEnteringVehicle())
```
Comment (theSarrum) [inline Client/mods/deathmatch/logic/CClientPed.cpp:2416]:
```suggestion
        else if (IsEnteringVehicle() && strcmp(szSimpleTaskName, "TASK_SIMPLE_GO_TO_POINT") != 0)
```
Did a lot of tests and the `TASK_SIMPLE_GO_TO_POINT` task in MTA is only performed when the player walks to a vehicle to enter it and this is what is needed to solve #1594.

File: Shared/mods/deathmatch/logic/Utils.cpp [hunk 3]
```diff
@@ -1134,3 +1134,23 @@ CVector ConvertEulerRotationOrder(const CVector& a_vRotation, eEulerRotationOrde
...
+
+bool definitelyGreaterThan(float a, float b, float epsilon)
+{
+    return (a - b) > ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * FLOAT_EPSILON);
+}
+
+bool definitelyLessThan(float a, float b, float epsilon)
+{
+    return (b - a) > ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * FLOAT_EPSILON);
+}
```
Comment (theSarrum) [inline Shared/mods/deathmatch/logic/Utils.cpp:1152]:
```suggestion
bool approximatelyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool essentiallyEqual(float a, float b, float epsilon)
{
    return fabs(a - b) <= ((fabs(a) > fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool definitelyGreaterThan(float a, float b, float epsilon)
{
    return (a - b) > ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}

bool definitelyLessThan(float a, float b, float epsilon)
{
    return (b - a) > ((fabs(a) < fabs(b) ? fabs(b) : fabs(a)) * epsilon);
}
```

**Review by theSarrum — CHANGES_REQUESTED** [review]
